src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1828,49 +1862,33 @@ final void registerWorker(WorkQueue w) {
      * @param ex the exception causing failure, or null if none
      */
     final void deregisterWorker(ForkJoinWorkerThread wt, Throwable ex) {
-        if ((runState & STOP) != 0L)       // ensure released
-            releaseAll();
-        WorkQueue w = null;
-        int src = 0, phase = 0;
-        boolean replaceable = false;
-        if (wt != null && (w = wt.workQueue) != null) {
-            phase = w.phase;
-            if ((src = w.source) != DROPPED) {
-                w.source = DROPPED;        // else already dropped
-                if (phase != 0) {          // else failed to start
-                    replaceable = true;
-                    if (w.top - w.base > 0) {
-                        ForkJoinTask<?> t; // cancel remaining tasks
-                        while ((t = w.nextLocalTask()) != null) {
-                            try {
-                                t.cancel(false);
-                            } catch (Throwable ignore) {
-                            }
-                        }
-                    }
-                }
-            }
-        }
-        if (src != DROPPED) {             // decrement counts
-            long c = ctl;
+        WorkQueue w; int phase;
+        if ((w = ((wt == null) ? null : wt.workQueue)) == null)
+            phase = 0;
+        else if ((phase = w.phase) != 0 && (phase & IDLE) != 0)
+            releaseWaiters();              // ensure released
+        if (w == null || w.source != DROPPED) {
```

Comment:
@DougLea If this isn't intended to be an `else-if` I'd recommend adding a line of whitespace between it and the line above.

Reply:
Thanks for the prod. While making the control flow here read better, I noticed that adding a leading runState check in registerWorker further reduced overhead a bit when shutdown occurs while still ramping up.